notebooks/05b Modelling - regression fairness.ipynb

```diff
@@ -0,0 +1,440 @@
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "## Load models"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1652797617528
+    },
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# load models from outside the git tree\n",
+    "with open(\"../../models/regression.pickle\", \"rb\") as handle:\n",
+    "    models = pickle.load(handle)\n",
+    "models"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "## Generate model predictions"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1652797619438
+    },
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "model = \"catboost\"\n",
+    "# Generate predictions\n",
+    "preds = np.clip(\n",
+    "    models[model][\"model\"].predict(X_test.drop(columns=sensitive_columns)), 0, None\n",
+    ")\n",
+    "\n",
+    "# Create a combined data frame\n",
+    "fairness_df = X_test.copy()\n",
+    "fairness_df[\"LENGTH_OF_STAY\"] = y_test\n",
+    "fairness_df[\"LENGTH_OF_STAY_PREDICTED\"] = preds"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "## Explore underlying distribution of sensitive features\n",
+    "\n",
+    "How many of each subcategory are represented in the data?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "gather": {
+     "logged": 1652798061346
+    },
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "for feature in sensitive_columns:\n",
+    "    data = pd.DataFrame(columns=[\"subfeature\", \"count\"])\n",
+    "    # iterate over all subcategories for the feature\n",
+    "    for subfeature in fairness_df[feature].unique():\n",
+    "        subset = fairness_df[fairness_df[feature] == subfeature]\n",
+    "        count = subset.LENGTH_OF_STAY.count()\n",
+    "        df = pd.DataFrame(data=[{\"subfeature\": subfeature, \"count\": count}])\n",
+    "        data = pd.concat([data, df])\n",
+    "\n",
+    "    fig = px.bar(data, x=\"subfeature\", y=\"count\", title=feature, orientation=\"v\")\n",
+    "    fig.add_hline(y=data[\"count\"].mean(), line_dash=\"dash\", line_color=\"green\")\n",
```

Comment:
Horizontal line only covers the first two bars in each of these plots, not sure if its supposed to appear like that

Reply:
Comment added ref: bug in Azure ML notebook rendering engine